README.md

# io.github.kare Java Library

## Development

Run all tests:

```bash
make test
```

Generate code coverage report of all sources:

```bash
make coverage
```


CHANGELOG.md
# Changelog

## v0.0.3-beta.15 (2021-09-06)

### Continuous Integration

- Add make recipe for GitHub git config (9b9d836)
- Tag matchers are globs not regex (492e6c3)

### Documentation

- Update RELEASE-NOTES.md for release 0.0.3-beta.15 (9f76eb4)
- Update CHANGELOG.md for release 0.0.3-beta.15 (1b0e431)

---

## v0.0.3-beta.14 (2021-09-06)

### Continuous Integration

- Fix semver regex (8227600)

### Documentation

- Update RELEASE-NOTES.md for release 0.0.3-beta.14 (3663d76)
- Update CHANGELOG.md for release 0.0.3-beta.14 (57125ee)

---

## v0.0.3-beta.13 (2021-09-06)

### Build

- Remove debug print (6df2ffc)
- Set project's basic attributes (bba5af5)

### Continuous Integration

- Create CHANGELOG (7dcce0e)
- Debug create release workflow (263ce0f)
- Print info on build (c6848ed)
- Do pre release on tag only (9013ffb)
- Install git-sv from GitHub Releases (86a38ba)
- Print env (99fc6d3)
- Fix typo in workflow (1e6a260)
- Fix typos (78b2bbf)
- Fix next version (6d2b7a5)
- Create Release from GitHub Actions UI (b7aff55)
- Fix missing bracket (66f7bca)
- Cache Gradle build (73887d4)
- Specify cache action version (cab6840)
- Fix typo in quotes (cca76d2)
- Remove spaces around the variabe definition (0997753)
- Add Gradle build cache (3b59c38)
- Set user account properties for publishing (fd10281)
- Change defaults for version inputs (b86dcdc)
- Fix quotes on gradle.properties (15efb5e)
- Remove redundant quotes (1550499)

### Documentation

- Update RELEASE-NOTES.md for release 0.0.3-beta.13 (14c565e)
- Update CHANGELOG.md for release 0.0.3-beta.13 (a0d2ece)

---

## v0.0.3-beta.12 (2021-09-05)

### Continuous Integration

- Remove redundant config (bab3aa3)
- Move Git config to env step (0f08f76)
- Add configure step to set Git account details (883ba06)
- Print release stacktrace on error (14f2738)
- Fix release notes in GitHub Release (4b0cd04)
- Update Create Release workflow (777b93d)

### Documentation

- Update CHANGELOG.md for release 0.0.3-beta.12 (d9ba366)

---

## v0.0.3-beta.11 (2021-09-05)

### Continuous Integration

- Debug print version (0dbdd06)

### Documentation

- Update CHANGELOG.md for release 0.0.3-beta.11 (afd54e8)

---

## v0.0.3-beta.10 (2021-09-05)

### Continuous Integration

- Validate release (c2af72e)
- Validate given version (7046503)
- Mark pre (alpha/beta) releases as such on GitHub Releases (ec1347c)
- Add Create Release workflow (eb0f9ea)
- Include release notes to GitHub Release body text (46c4237)

### Documentation

- Update CHANGELOG.md for release 0.0.3-beta.10 (70a96d2)

---

## v0.0.3-beta.9 (2021-09-05)

### Continuous Integration

- Remove Bash arguments (e118202)

### Documentation

- Update CHANGELOG.md for release 0.0.3-beta.9 (e34ce7d)

---

## v0.0.3-beta.8 (2021-09-05)

### Build

- Add sv4git configuration (9c46bd5)
- Don't use profile or rc for Bash init (944283b)

### Continuous Integration

- Set CI Bash arguments (19676da)
- Publish pre releases to Maven Central (4d0a777)
- Set Bash flags (1305419)
- Define Java version by major release (bb86ef2)
- Try github.com/bvieira/sv4git for automatic release notes (714f50d)

### Documentation

- Update CHANGELOG.md for release 0.0.3-beta.8 (556731e)

---

## v0.0.3-beta.7 (2021-09-03)

### Continuous Integration

- Debug output (d9310ca)

---

## v0.0.3-beta.6 (2021-09-02)

### Continuous Integration

- Fix empty release notes file (edc62c2)
- Enable sv4git (bdeee06)

---

## v0.0.3-beta.5 (2021-09-02)

### Continuous Integration

- Fix build error by removing empty line (f505bf7)

---

## v0.0.3-beta.4 (2021-09-02)

---

## v0.0.3-beta.3 (2021-09-02)

### Continuous Integration

- Disable sv4git (817b543)
- Debug release notes generation (c2c1bec)
- Set Gradle build '--console auto' (e26f594)

---

## v0.0.3-beta.2 (2021-09-02)

### Build

- Move version to gradle.properties (61d701e)
- Move group to gradle.properties (a9471b3)

### Continuous Integration

- **deps:** bump actions/setup-go from 2.1.3 to 2.1.4 (#3) (49c21d1)
- Add Git checkout to Pre Release (f164064)
- Define Go version for Pre Release (15fa86e)
- Fix pre release workflow (e9376c8)

---

## v0.0.3-beta.1 (2021-08-24)

### Features

- **common:** Add IO.copy(java.io.InputStream, java.io.OutputStream) (ab295d4)
- **common:** Add IO.close(java.io.Closeable...) (8ed4578)
- **common:** Add ByteArrays.cat(byte[], byte[]) (b1d2c3d)
- **common:** Add ByteArrays.contains(byte[], byte) (1d17fad)

### Testing

- Update JUnit Jupiter 5.7.2 (b9ad544)
- Use full byte range -128..127 while testing ByteArrays.reverse() (f8ebb05)

### Build

- Update to Gradle 7.2 (44e87c5)
- **deps:** bump com.diffplug.spotless from 5.12.5 to 5.14.3 (#1) (ba13eb3)
- **deps:** Update google-java-format 1.11.0 (9841bb7)
- Use Java 8 when generating Javadoc (996c780)
- Compile with Java 8 (c6e1465)
- Run unit tests in parallel (ecf800f)
- Add issue, CI, license and organization links to Maven publication (9ec293e)
- Define Gradle runtime and optimize build with parallel, cache and watch (cf43904)
- Make build logging more verbose (d1c6404)
- Set common Javadoc build options (48d7192)
- Build project wide Javadoc (6e02365)
- Use private key to sign build artifacts (6159638)
- Skip signing SNAPSHOTs and with skipSigning Gradle property (d1595b3)
- Use console=verbose as build output level (5a03edb)

### Continuous Integration

- Pre Release on Tag (c27db49)
- **deps:** bump actions/setup-java from 2.2.0 to 2.3.0 (#2) (da8d165)
- Use Java 11 to run build (16674d4)
- Add basic continuous integration build (4cf3f90)
- Add Dependabot for Gradle and GitHub Action dependencies (f270ee9)

### Documentation

- Update ByteArrays.toLong() Javadoc (e5a6a4d)
- Format README code examples (7109b7c)

---

## v0.0.2 (2021-06-02)

### Features

- ByteArrays.reverse(byte[]) (3f2f244)
- **build:** add more recipes (f519cb9)

### Build

- **java-lib:** fail on dependency version conflict (fc5def7)

---

## v0.0.1 (2021-05-25)

---

## v0.0.1-beta.1 (2021-05-25)

---



common/src/test/java/io/github/kare/common/IOTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package io.github.kare.common;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.Closeable;
import java.io.IOException;
import java.io.OutputStream;
import org.junit.jupiter.api.Test;

class IOTest {
  @Test
  void copyCopiesBytes() throws IOException {
    final byte[] expected = "Hello, World!".getBytes();
    OutputStream out = new ByteArrayOutputStream();
    long count = IO.copy(new ByteArrayInputStream(expected), out);
    assertArrayEquals(expected, out.toString().getBytes());
    assertEquals(expected.length, count);
  }

  @Test
  void closeClosesCloseables() {
    class BasicCloseable implements Closeable {
      boolean isCalled = false;

      @Override
      public void close() {
        isCalled = true;
      }
    }
    class ThrowingCloseable implements Closeable {
      boolean isCalled = false;

      @Override
      public void close() throws IOException {
        isCalled = true;
        throw new IOException("Mock class representing IO error");
      }
    }

    BasicCloseable c1 = new BasicCloseable();
    IO.close(c1);
    assertTrue(c1.isCalled);

    ThrowingCloseable t1 = new ThrowingCloseable();
    IO.close(t1);
    assertTrue(t1.isCalled);

    ThrowingCloseable t2 = new ThrowingCloseable();
    BasicCloseable c2 = new BasicCloseable();
    ThrowingCloseable t3 = new ThrowingCloseable();
    BasicCloseable c3 = new BasicCloseable();
    IO.close(null, t2, c2, null, t3, c3, null);
    assertTrue(t2.isCalled);
    assertTrue(c2.isCalled);
    assertTrue(t3.isCalled);
    assertTrue(c3.isCalled);
  }

  @Test
  void closeWithNoArgs() {
    IO.close();
    assertTrue(true);
  }

  @Test
  void closeWithNullCloseableArg() {
    Closeable closeable = null;
    IO.close(closeable);
    assertTrue(true);
  }

  @Test
  void closeThrowsWithNullCloseablesArray() {
    Closeable[] closeables = null;
    try {
      IO.close(closeables);
      fail("Excepting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }
  }
}


common/src/test/java/io/github/kare/common/ByteArraysTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package io.github.kare.common;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import java.math.BigInteger;
import org.junit.jupiter.api.Test;

class ByteArraysTest {
  @Test
  void toLongByteArrayMustBeEightBytes() {
    try {
      ByteArrays.toLong(new byte[2]);
      fail("expecting ArrayIndexOutOfBoundsException");
    } catch (ArrayIndexOutOfBoundsException e) {
      assertTrue(true);
    }
  }

  @Test
  void toLongByteArrayMustNotBeNull() {
    try {
      ByteArrays.toLong(null);
      fail("expecting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }
  }

  @Test
  void toLongConvertsEightByteArrayToLong() {
    byte[] expected = {
      1, 2, 3, 4, 5, 6, 7, 8,
    };
    long result = ByteArrays.toLong(expected);
    assertEquals(72623859790382856L, result);
    byte[] bytes = new BigInteger(String.valueOf(result)).toByteArray();
    assertArrayEquals(expected, bytes);
  }

  @Test
  void reverseByteArray() {
    byte[] expected = {1};
    byte[] given = {1};
    ByteArrays.reverse(given);
    assertArrayEquals(expected, given);

    expected = new byte[] {1, 2, 3, 4, 5, 6, 7, 8};
    given = new byte[] {8, 7, 6, 5, 4, 3, 2, 1};
    ByteArrays.reverse(given);
    assertArrayEquals(expected, given);

    expected =
        new byte[] {
          127, 126, 125, 124, 123, 122, 121, 120, 119, 118, 117, 116, 115, 114, 113, 112, 111, 110,
          109, 108, 107, 106, 105, 104, 103, 102, 101, 100, 99, 98, 97, 96, 95, 94, 93, 92, 91, 90,
          89, 88, 87, 86, 85, 84, 83, 82, 81, 80, 79, 78, 77, 76, 75, 74, 73, 72, 71, 70, 69, 68,
          67, 66, 65, 64, 63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 47, 46,
          45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24,
          23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1,
          -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -13, -14, -15, -16, -17, -18, -19, -20,
          -21, -22, -23, -24, -25, -26, -27, -28, -29, -30, -31, -32, -33, -34, -35, -36, -37, -38,
          -39, -40, -41, -42, -43, -44, -45, -46, -47, -48, -49, -50, -51, -52, -53, -54, -55, -56,
          -57, -58, -59, -60, -61, -62, -63, -64, -65, -66, -67, -68, -69, -70, -71, -72, -73, -74,
          -75, -76, -77, -78, -79, -80, -81, -82, -83, -84, -85, -86, -87, -88, -89, -90, -91, -92,
          -93, -94, -95, -96, -97, -98, -99, -100, -101, -102, -103, -104, -105, -106, -107, -108,
          -109, -110, -111, -112, -113, -114, -115, -116, -117, -118, -119, -120, -121, -122, -123,
          -124, -125, -126, -127, -128
        };
    given =
        new byte[] {
          -128, -127, -126, -125, -124, -123, -122, -121, -120, -119, -118, -117, -116, -115, -114,
          -113, -112, -111, -110, -109, -108, -107, -106, -105, -104, -103, -102, -101, -100, -99,
          -98, -97, -96, -95, -94, -93, -92, -91, -90, -89, -88, -87, -86, -85, -84, -83, -82, -81,
          -80, -79, -78, -77, -76, -75, -74, -73, -72, -71, -70, -69, -68, -67, -66, -65, -64, -63,
          -62, -61, -60, -59, -58, -57, -56, -55, -54, -53, -52, -51, -50, -49, -48, -47, -46, -45,
          -44, -43, -42, -41, -40, -39, -38, -37, -36, -35, -34, -33, -32, -31, -30, -29, -28, -27,
          -26, -25, -24, -23, -22, -21, -20, -19, -18, -17, -16, -15, -14, -13, -12, -11, -10, -9,
          -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
          17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
          39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60,
          61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82,
          83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103,
          104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121,
          122, 123, 124, 125, 126, 127,
        };
    ByteArrays.reverse(given);
    assertArrayEquals(expected, given);
  }

  @Test
  void reverseByteArrayMustNotBeNull() {
    try {
      ByteArrays.reverse(null);
      fail("expecting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }
  }

  @Test
  void containsByteArrayMustNotBeNull() {
    try {
      ByteArrays.contains(null, (byte) 1);
      fail("expecting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }
  }

  @Test
  void containsByteArray() {
    byte[] given = new byte[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    assertTrue(ByteArrays.contains(given, (byte) 0));
    assertTrue(ByteArrays.contains(given, (byte) 5));
    assertTrue(ByteArrays.contains(given, (byte) 10));
    assertFalse(ByteArrays.contains(given, (byte) 11));
    assertFalse(ByteArrays.contains(new byte[] {}, (byte) 11));
  }

  @Test
  void catNullArguments() {
    try {
      ByteArrays.cat(null, null);
      fail("expecting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }

    try {
      ByteArrays.cat(null, new byte[] {});
      fail("expecting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }

    try {
      ByteArrays.cat(new byte[] {}, null);
      fail("expecting NullPointerException");
    } catch (NullPointerException e) {
      assertTrue(true);
    }
  }

  @Test
  void catConcatenatesTwoArrays() {
    byte[] a = new byte[] {1};
    byte[] b = new byte[] {2};
    byte[] result = ByteArrays.cat(a, b);
    assertArrayEquals(new byte[] {1, 2}, result);

    a = new byte[] {1, 2, 3, 4, 5, 6, 7};
    b = new byte[] {-2, -1, 0, 1};
    result = ByteArrays.cat(a, b);
    assertArrayEquals(new byte[] {1, 2, 3, 4, 5, 6, 7, -2, -1, 0, 1}, result);
  }
}


common/src/main/java/io/github/kare/common/IO.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package io.github.kare.common;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Common IO utilities. */
public final class IO {
  /** Copy buffer size in bytes. */
  private static final int COPY_BUFFER_SIZE = 1024 * 4;

  private IO() {
    throw new AssertionError("No IO instances for you!");
  }

  /**
   * Copy bytes from input stream {@code in} to output stream {@code out}.
   *
   * @param in Input stream to read bytes from
   * @param out Output stream to write bytes to
   * @return Number of bytes copies
   * @throws IOException Thrown in case of I/O error
   * @throws NullPointerException If either given in or out is null
   */
  public static long copy(final InputStream in, final OutputStream out) throws IOException {
    final byte[] buf = new byte[COPY_BUFFER_SIZE];
    int n;
    long byteCount = 0;
    while ((n = in.read(buf, 0, buf.length)) != -1) {
      out.write(buf, 0, n);
      byteCount += n;
    }
    return byteCount;
  }

  /**
   * Close given Closeables unless they're null.
   *
   * @param closeables Closeable instances to call close()
   * @throws NullPointerException If given closeables array is null
   */
  public static void close(final Closeable... closeables) {
    for (final Closeable closeable : closeables) {
      if (closeable != null) {
        try {
          closeable.close();
        } catch (final IOException e) {
          Logger logger = Logger.getLogger(IO.class.getName());
          logger.log(Level.SEVERE, "Error while closing I/O resource", e);
        }
      }
    }
  }
}


common/src/main/java/io/github/kare/common/ByteArrays.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package io.github.kare.common;

/** Common utilities that operate on {@code byte} arrays. */
public final class ByteArrays {
  private ByteArrays() {
    throw new AssertionError("No ByteArrays instances for you!");
  }

  /**
   * Convert an array of bytes to long. Only first eight bytes are considered.
   *
   * @param bytes Array of eight bytes
   * @return Returns given byte array as long
   * @throws ArrayIndexOutOfBoundsException Thrown if given array is not minimum of eight bytes in
   *     length
   * @throws NullPointerException Thrown if given array is null
   */
  public static long toLong(final byte[] bytes) {
    return ((long) (bytes[0]) << 56)
        + (((long) bytes[1] & 0xFF) << 48)
        + ((long) (bytes[2] & 0xFF) << 40)
        + ((long) (bytes[3] & 0xFF) << 32)
        + ((long) (bytes[4] & 0xFF) << 24)
        + ((bytes[5] & 0xFF) << 16)
        + ((bytes[6] & 0xFF) << 8)
        + (bytes[7] & 0xFF);
  }

  /**
   * Reverses given byte array.
   *
   * @param bytes Array of bytes
   * @throws NullPointerException Thrown if given array is null
   */
  public static void reverse(byte[] bytes) {
    for (int i = 0; i < bytes.length / 2; i++) {
      final int n = bytes.length - 1 - i;
      bytes[i] = (byte) ((int) bytes[i] ^ (int) bytes[n]);
      bytes[n] = (byte) ((int) bytes[i] ^ (int) bytes[n]);
      bytes[i] = (byte) ((int) bytes[i] ^ (int) bytes[n]);
    }
  }

  /**
   * Checks whether given {@code bytes} array contains value {@code b}.
   *
   * @param bytes Byte array to search
   * @param b byte to search in array bytes
   * @return true if given bytes contains b, false otherwise
   * @throws NullPointerException Thrown if given array is null
   */
  public static boolean contains(final byte[] bytes, final byte b) {
    for (final byte a : bytes) {
      if (a == b) {
        return true;
      }
    }
    return false;
  }

  /**
   * Concatenates two byte arrays to a single array. Reserves memory for a new array.
   *
   * @param a Start of the new array
   * @param b Bytes after array a
   * @return Array a + b concatenated
   * @throws NullPointerException Thrown if any given array is null
   */
  public static byte[] cat(final byte[] a, final byte[] b) {
    final int aLen = a.length;
    final byte[] result = new byte[aLen + b.length];
    System.arraycopy(a, 0, result, 0, aLen);
    System.arraycopy(b, 0, result, aLen, b.length);
    return result;
  }
}
